### DISCUSSION TITLE
Mixin-with-prefix capability?

### DISCUSSION THREAD

**Discussion by rcauble** [remkop/picocli#1808] [Q&A] [answered]
Hi,

I have a use case where I have 2 commands with each their own set of settings:

```
class Command1Settings {
    @CommandLine.Option(names = "--foo")
    private String foo;
    @CommandLine.Option(names = "--bar")
    private String bar;
}

@CommandLine.Command(name = "command1")
public static final class Command1  {
    @CommandLine.Mixin
    private Command1Settings settings = new Command1Settings();

    public void run() {
        String results1 = Command1Runner.run(settings);
        System.out.println(results1);
    }
}

class Command2Settings {
    @CommandLine.Option(names = "--foo")
    private String foo;
    @CommandLine.Option(names = "--bar")
    private String bar;
}

@CommandLine.Command(name = "command2")
 public static final class Command2  {
    @CommandLine.Mixin
    private Command2Settings settings = new Command2Settings();
    public void run() {
        String results2 = Command2Runner.run(settings);
         System.out.println(results2);
    }
}
```

And I am trying to create a 3rd command that computes a diff of the other 2. As you can see above, though there is overlap in the options of the 2 commands and so I'm looking for a MixinWithPrefix capability so that I can do something like this:

```
@CommandLine.Command(name = "diff")
public static final class Diff  {
    @CommandLine.MixinWithPrefix("first-")
    private Command1Settings settings1 = new Command1Settings();
    @CommandLine.MixinWithPrefix("second-")
    private Command1Settings settings2 = new Command2Settings();
    public void run() {
        String results1 = Command1Runner.run(settings1);
        String results2 = Command2Runner.run(settings2);
         System.out.println(diff(results1, results2));
    }
}
```

And then this will define a new command where the user can do:

```
diff --first-foo thing1 --first-bar thing2 --second-foo thing3 --second-bar thing4
```

To add an additional constraint for context, Command1 and Command2 live in different libraries from the differ and so I can't simply arrange for the names to be distinct as they are not necessarily owned by me.

Any suggestions?

**Comment by remkop** [comment, marked as answer]
You may be able to use [variable interpolation](https://picocli.info/#_variable_interpolation) and [resource bundles](https://picocli.info/#_internationalization) to give options in one command a different name than in another command.
The options need to be defined something like this:

```java
    @Option(names = "${lookupkey.foo:---foo") // yes 3 dashes, that is not a typo, the ":-" means "what follows is the default"
    private String foo;
    @Option(names = "${lookupkey.bar:---bar")
    private String bar;
```

Then, in your diff application, you add a resource bundle that provides the option name.
See this example: https://picocli.info/#_shared_resource_bundles

In the resource bundle, you prefix the `lookupkey.foo` with the command name so you can have separate values for each command.

**Reply by remkop** [reply]
@rcauble Did this answer your question?

**Reply by rcauble** [reply]
Thanks, yeah. The one drawback is that I don't necessarily know the full list of options in each of the sub-modules as they are owned by a different team and may change over time without anyone telling me to update the resource bundle. I had started to look at IModelTransformer as a possible way to do this but got pulled away to something else. Would you suggest that as a possible avenue?

**Comment by reddaly** [comment]
I'm not sure the proposed solution fixed my particular use case.

I'd like a mixin that allows specifying a few flags for outputting spreadsheet-like information. One flag is needed for the CSV output location and another is needed for the XSLT location. The same command has multiple spreadsheets being output.

e.g.

```shell
process-results --input /foo/bar.pbbin --output_costs_csv costs.csv --output_costs_google_sheet "https://docs.google.com/spreadsheets/d/dlfkasdlnkldsanlkfasdlkj" --output_stats_csv /tmp/stats.csv
```

If we could change a variable per-mixin-instance, the above would work. But it seems all the mixins used by a command would share the same set of variables.

**Reply by remkop** [reply]
@reddaly Yes, that sounds like a different use case. Would you mind creating a separate ticket/discussion topic?  (And can you provide a bit more detail on what is not working as expected)
